rustc_lint: Enforce `rustc::potential_query_instability` lint

Enselic opened the pull request:
Stop allowing `rustc::potential_query_instability` on all of `rustc_lint` and instead allow it on a case-by-case basis if it is safe to do so. In this particular crate, all lints were safe to allow.

Part of https://github.com/rust-lang/rust/issues/84447 which is E-help-wanted.

Enselic committed:
rustc_lint: Enforce `rustc::potential_query_instability` lint

Stop allowing `rustc::potential_query_instability` on all of
`rustc_lint` and instead allow it on a case-by-case basis if it is safe
to do so. In this particular crate, all lints were safe to allow.

cjgillot commented on compiler/rustc_lint/src/levels.rs:
```diff
@@ -1028,6 +1028,8 @@ impl<'s, P: LintLevelsProvider> LintLevelsBuilder<'s, P> {
         }
 
         if !is_crate_node {
+            // We break on the first lint, and it does not matter which one.
```
It may not matter, but emitting a different diagnostic depending on the architecture is a bug.

Enselic replied:
I changed `current_specs()` to return a `FxIndexMap`, which only cascaded to 13 other places, which seems fine. So now the iteration order here should be deterministic.

(Due to conflicts I also had to rebase.)

bors commented:
:umbrella: The latest upstream changes (presumably #119421) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Enselic committed:
rustc_lint: Make `LintLevelsProvider::current_specs()` return `&FxIndexMap`

So that lint iteration order becomes predicitable. Discovered with
`rustc::potential_query_instability`.

Enselic commented:
r? @cjgillot

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 295d6003acaa8da531248422a9d8567e30aeb52a has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 295d6003acaa8da531248422a9d8567e30aeb52a -->
<!-- homu: {"type":"Approved","sha":"295d6003acaa8da531248422a9d8567e30aeb52a","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 295d6003acaa8da531248422a9d8567e30aeb52a with merge e51e98dde6a60637b6a71b8105245b629ac3fe77...
<!-- homu: {"type":"BuildStarted","head_sha":"295d6003acaa8da531248422a9d8567e30aeb52a","merge_sha":"e51e98dde6a60637b6a71b8105245b629ac3fe77"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7371854518/job/20060580630)
Approved by: cjgillot
Pushing e51e98dde6a60637b6a71b8105245b629ac3fe77 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7371854518/job/20060580630"},"merge_sha":"e51e98dde6a60637b6a71b8105245b629ac3fe77"} -->

rust-timer commented:
Finished benchmarking commit (e51e98dde6a60637b6a71b8105245b629ac3fe77): [comparison URL](https://perf.rust-lang.org/compare.html?start=67b6975051b83ef2bd28f06e8467470d570aceb3&end=e51e98dde6a60637b6a71b8105245b629ac3fe77&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=67b6975051b83ef2bd28f06e8467470d570aceb3&end=e51e98dde6a60637b6a71b8105245b629ac3fe77&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.2%, 0.3%] | 2     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.2%, 0.4%] | 8     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.2%, 0.3%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=67b6975051b83ef2bd28f06e8467470d570aceb3&end=e51e98dde6a60637b6a71b8105245b629ac3fe77&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-2.2%, -0.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.3% | [-2.2%, -0.4%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=67b6975051b83ef2bd28f06e8467470d570aceb3&end=e51e98dde6a60637b6a71b8105245b629ac3fe77&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=67b6975051b83ef2bd28f06e8467470d570aceb3&end=e51e98dde6a60637b6a71b8105245b629ac3fe77&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 665.69s -> 666.95s (0.19%)
**Artifact size**: 311.76 MiB -> 311.74 MiB (-0.01%)

Enselic commented:
The performance regression should be fixed by https://github.com/rust-lang/rust/pull/119488.
